helix-rest/src/main/java/org/apache/helix/rest/metadatastore/MetadataStoreDirectory.java

```diff
@@ -0,0 +1,134 @@
+   */
+  Collection<String> getAllNamespaces();
+
+  /**
+   * Returns all metadata store realms in the given namespace.
+   * @return
+   */
+  Collection<String> getAllMetadataStoreRealms(String namespace);
+
+  /**
+   * Returns all path-based sharding keys in the given namespace.
+   * @return
+   */
+  Collection<String> getAllShardingKeys(String namespace);
+
+  /**
+   * Returns all path-based sharding keys in the given namespace and the realm.
+   * @param namespace
+   * @param realm
+   * @return
+   */
+  Collection<String> getAllShardingKeysInRealm(String namespace, String realm);
+
+  /**
+   * Returns all sharding keys that have the given path as the prefix substring.
+   * E.g) Given that there are sharding keys: /a/b/c, /a/b/d, /a/e,
+   * getAllShardingKeysUnderPath(namespace, "/a/b") returns ["/a/b/c": "realm", "/a/b/d": "realm].
+   * @param namespace
+   * @param path
+   * @return
+   */
+  Map<String, String> getAllMappingUnderPath(String namespace, String path);
+
+  /**
+   * Returns the name of the metadata store realm based on the namespace and the sharding key given.
+   * @param namespace
+   * @param shardingKey
+   * @return
+   */
+  String getMetadataStoreRealm(String namespace, String shardingKey);
+
+  /**
+   * NOTE: The following CRUD methods are idempotent.
+   */
+  /**
+   * Creates a namespace in the routing metadata store.
+   * @param namespace
+   * @return true if successful or if the namespace already exists. false if not successful.
+   */
+  boolean addNamespace(String namespace);
+
+  /**
+   * Deletes a namespace in the routing metadata store.
+   * @param namespace
+   * @return true if successful or if the namespace does not exist. false otherwise.
+   */
+  boolean deleteNamespace(String namespace);
+
+  /**
+   * Creates a realm. If the namespace does not exist, it creates one.
+   * @param namespace
+   * @param realm
+   * @return true if successful or if the realm already exists. false otherwise.
+   */
+  boolean addMetadataStoreRealm(String namespace, String realm);
+
+  /**
+   * Deletes a realm.
+   * @param namespace
+   * @param realm
+   * @return true if successful or the realm or namespace does not exist. false otherwise.
+   */
+  boolean deleteMetadataStoreRealm(String namespace, String realm);
+
+  /**
+   * Creates a mapping between the sharding key to the realm in the given namespace.
+   * @param namespace
+   * @param realm
+   * @param shardingKey
+   * @return false if failed
+   */
+  boolean addShardingKey(String namespace, String realm, String shardingKey);
+
+  /**
+   * Deletes the mapping between the sharding key to the realm in the given namespace.
+   * @param namespace
+   * @param realm
+   * @param shardingKey
+   * @return false if failed; true if the deletion is successful or the key does not exist.
+   */
+  boolean deleteShardingKey(String namespace, String realm, String shardingKey);
+
+  /**
+   * Close MetadataStoreDirectory.
+   */
+  void close();
```

Comment:
Do we consider close operation is idempotent? If yes, we can think about extend the interface AutoCloseable. Then user can use try-with-resources.

Reply:
Using AutoCloseable could be a good idea. I'll consider adding it.